### PR TITLE
[DOCS] Introduction to Relay IR.

### PR DESCRIPTION
**Pull request by tqchen**
This is an introduction material to Relay IR for developers who have a background on data-flow and computational graphs. This tutorial is a result of discussion with @jroesch @yzhliu @junrushao1994 @MarisaKirisame @slyubomirsky @joshpoll and other folks. We try to blend the views from deep learning frameworks and PL

### CODE REVIEW

**Comment by junrushao1994** [comment]
cc @szha

File: docs/dev/relay_intro.rst [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+Hopefully, by now you are familiar with the two kinds of representations.
+Most functional programming languages do their analysis in A-normal form, in the case of A-normal form,
+the analyzer does not need to be mindful that the expressions are DAGs.
+
+Relay choose to support both the data-flow form and let binding. We believe that it is important to let the
+framework developer choose the representation they are familiar with.
+This does, however, have some implications on how we write passes:
+
+- If you come from a data-flow background and want to handle let, keep a map of var to the expressions so you can perform lookup when encountering a var. This is a likely means a minimum change as we already need a map from expr-> transformed expression anyway. Note that this will effectively remove all the let in the program.
```
Comment (masahi) [inline docs/dev/relay_intro.rst]: This is a likely means -> This likely means

File: docs/dev/relay_intro.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
+Introduction to Relay IR
+========================
+This article introduces Relay IR -- the second generation of NNVM.
+We expect readers from two kinds of background -- those who have a programming language background and deep learning
+framework developers who are familiar with the computational graph representation.
+This article is mainly written for deep learning framework developers who are familiar with the computational graph representation.
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: Is this a little repetitive?

File: docs/dev/relay_intro.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+We briefly summarize the design goal here, and will touch upon these points in the later part of the article.
+
+- Support traditional data flow style programming and transformations.
+- Support functional style scoping, let-binding and making it fully featured differentiable language.
+- Being able to allow the user to mix the two programming styles.
+
+Build Computational Graph with Relay
+------------------------------------
+Traditional deep learning frameworks use computational graphs as their intermediate representation.
+A computational graph (or data-flow graph), is a directed acyclic graph (DAG) that represent the computation.
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: represents

File: docs/dev/relay_intro.rst [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+.. image:: https://raw.githubusercontent.com/tvmai/tvmai.github.io/master/images/relay/dataflow.png
+    :align: center
+    :scale: 70%
+
+
+You can use Relay to build a computational(dataflow) graph. Specifically, the above code shows how to
+construct a simple two-node graph. You can find that the syntax of the example is not that different from existing
+computational graph IR like NNVMv1, with the only difference in terms of terminology:
+
+- Existing frameworks usually uses graph and subgraph
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: s/uses/use

File: docs/dev/relay_intro.rst [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+- Relay uses function e.g. --  ``fn (%x)``, to indicate the graph
+
+Each data-flow node is a CallNode in Relay. The relay python DSL allows you to construct a data-flow quickly.
+One thing we want to highlight in the above code -- is that we explicitly constructed an Add node with
+both input points to ``%1``.  When a deep learning framework evaluates the above program, it will compute
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: both input point to ?

File: docs/dev/relay_intro.rst [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  fn (%x) {
+    %1 = log(%x)
+    %2 = fn(%y) {
+      add(%y, %1)
+    }
+    %2
+  }
+
+Let binding solves this problem, as the computation of the value happens at the let node. In both programs,
+if we change ``%1 = log(%x)`` to ``let %v1 = log(%x)``, we clearly specifies the computation location to
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: s/specifies/specify

File: docs/dev/relay_intro.rst [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+be outside of the if scope and closure. As you can see let-binding gives a more precise specification of the computation site
+and could be useful when we generate backend code(as such specification is in the IR).
+
+On the other hand, the data-flow form, which does not specify the scope of computation, does have its own advantages
+-- we don’t need to worry about where to put the let when we generate the code. The dataflow form also gives more freedom
+to the later passes to decide where to put the evaluation point. As a result, it might not be a bad idea to use data flow
+form of the program in the initial phases of optimizations when you find it is convenient.
+As a matter of fact, many optimizations in relay today are written to optimize dataflow programs.
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: s/relay/Relay ?

**Comment by tqchen** [comment]
Thanks @masahi @zhiics for comments, I have updated accordingly

File: docs/dev/relay_intro.rst [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+
+The nested let-binding is called A-normal form, and it is commonly used as IRs in functional programming languages.
+Now, please take a close look at the AST structure. While the two programs are semantically identical
+(so are their textual representations, except that A-normal form has let prefix), their AST structures are different from each other.
+
+Since program optimizations take these AST data structures and transform them, the two different structure will
+affect the compiler code we are going to write. For example, if we want to detect a pattern ``add(log(x), y)``:
+
+- In the data-flow form, we can first access the add node, then directly look at its first arguments to see if it is a log
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: s/arguments/argument

File: docs/dev/relay_intro.rst [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+computational graph IR like NNVMv1, with the only difference in terms of terminology:
+
+- Existing frameworks usually use graph and subgraph
+- Relay uses function e.g. --  ``fn (%x)``, to indicate the graph
+
+Each data-flow node is a CallNode in Relay. The relay python DSL allows you to construct a data-flow quickly.
+One thing we want to highlight in the above code -- is that we explicitly constructed an Add node with
+both input point to ``%1``.  When a deep learning framework evaluates the above program, it will compute
+the nodes in topological order, and ``%1`` will only be computed once.
+While the this fact is very natural to deep learning framework builders, it is something that might
```
Comment (zhiics) [inline docs/dev/relay_intro.rst]: While this fact

**Comment by zhiics** [comment]
I found two more typos after another reading.

**Comment by tqchen** [comment]
Thanks @masahi @zhiics this is merged

**Pull request merged**